pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go

```diff
@@ -0,0 +1,458 @@
+	}
+	var res []model.AddressInfo
+	var removed []string
+	got := sets.New[string]()
+	for wname := range addresses {
+		wl := a.Lookup(wname)
+		if len(wl) == 0 {
+			removed = append(removed, wname)
+		} else {
+			for _, addr := range wl {
+				if !got.InsertContains(addr.ResourceName()) {
+					res = append(res, addr)
+				}
+			}
+		}
+	}
+	return res, sets.New(removed...)
+}
+
+func (a *index) WorkloadsForWaypoint(scope model.WaypointScope) []model.WorkloadInfo {
+	// Lookup scope. If its namespace wide, remove entries that are in SA scope
+	workloads := a.workloads.ByOwningWaypoint.Lookup(scope)
+
+	if scope.ServiceAccount == "" {
+		// This is a namespace wide waypoint. Per-SA waypoints have precedence, so we need to filter them out
+		workloads = slices.FilterInPlace(workloads, func(info model.WorkloadInfo) bool {
+			s := model.WaypointScope{
+				Namespace:      info.Namespace,
+				ServiceAccount: info.ServiceAccount,
+			}
+			// If there is a more specific waypoint, skip this
+			return len(a.waypoints.ByScope.Lookup(s)) == 0
+		})
+	}
+	return workloads
+}
+
+// Waypoint finds all waypoint IP addresses for a given scope.  Performs first a Namespace+ServiceAccount
+// then falls back to any Namespace wide waypoints
+func (a *index) Waypoint(scope model.WaypointScope) []netip.Addr {
+	res := sets.Set[netip.Addr]{}
+	waypoints := a.waypoints.ByScope.Lookup(scope)
+	if len(waypoints) == 0 {
+		// Now look for namespace-wide
+		scope.ServiceAccount = ""
+		waypoints = a.waypoints.ByScope.Lookup(scope)
+	}
+	for _, waypoint := range waypoints {
+		res.Insert(waypoint.Addresses[0])
+	}
+	return res.UnsortedList()
+}
+
+func (a *index) AdditionalPodSubscriptions(
+	proxy *model.Proxy,
+	allAddresses sets.String,
+	currentSubs sets.String,
+) sets.String {
+	shouldSubscribe := sets.New[string]()
+
+	// First, we want to handle VIP subscriptions. Example:
+	// Client subscribes to VIP1. Pod1, part of VIP1, is sent.
+	// The client wouldn't be explicitly subscribed to Pod1, so it would normally ignore it.
+	// Since it is a part of VIP1 which we are subscribe to, add it to the subscriptions
+	for addr := range allAddresses {
+		for _, wl := range model.ExtractWorkloadsFromAddresses(a.Lookup(addr)) {
+			// We may have gotten an update for Pod, but are subscribed to a Service.
+			// We need to force a subscription on the Pod as well
+			for namespacedHostname := range wl.Services {
+				if currentSubs.Contains(namespacedHostname) {
+					shouldSubscribe.Insert(wl.ResourceName())
+					break
+				}
+			}
+		}
+	}
+
+	// Next, as an optimization, we will send all node-local endpoints
+	if nodeName := proxy.Metadata.NodeName; nodeName != "" {
+		for _, wl := range model.ExtractWorkloadsFromAddresses(a.All()) {
+			if wl.Node == nodeName {
+				n := wl.ResourceName()
+				if currentSubs.Contains(n) {
+					continue
+				}
+				shouldSubscribe.Insert(n)
+			}
+		}
+	}
+
+	return shouldSubscribe
+}
+
+func (a *index) SyncAll() {
```

Comment:
Couldn't we make `NetworkManager` into a singleton that depends on Namespaces and MeshConfig/MeshNetworks? It's a decent amount of code, but would help hold off introducing manual recomputation triggers.

Reply:
I'd be happy to do this as a follow-up in addition to setting workloads' network gateway field

Reply:
Yeah I was just trying to make this PR not absolutely massive and not have 2 codepaths for network (or 1, but with stable functionality tied to krt -- right now krt only touches alpha code)